text/trace/0157-messaging-semantic-conventions.md

```diff
@@ -0,0 +1,184 @@
+| An OTEP proposing a set of attributes and conventions covering the scenarios in this document is approved and merged. | 10/29/2021 |
+| Proposed specification changes are verified by prototypes for the scenarios and examples below. | 11/15/2021 |
+| The [specification for messaging semantic conventions for tracing](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/messaging.md) are updated according to the OTEP mentioned above and are declared [stable](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/versioning-and-stability.md#stable). | 11/30/2021 |
+
+## Terminology
+
+### Message
+
+A "message" is a transport wrapper for the transfer of information between two
+or more parties. The information is a combination of data and metadata.
+Messages may be transferred directly between parties or via one or more
+intermediaries.  A message may carry annotations that are directed at
+intermediaries on the message path. Messages are uniquely identifiable.
+
+In the strict sense, a _message_ is an item of data that is sent to a specific
+destination, whereas an _event_ is a signal emitted by a component upon
+reaching a given state. This document is agnostic of those differences and uses
+the term "message" in a wider sense to cover both concepts.
+
+### Producer
+
+The "producer" is a specific instance, process or device that creates and
+publishes a message. "Publishing" is the process of sending a message or batch
+to the intermediary or consumer.
+
+### Consumer
+
+A "consumer" receives the message and acts upon it. It uses the context and
+data to execute some logic, which might lead to the occurrence of new events.
+
+The consumer receives, processes, and settles a message. "Receiving" is the
+process of obtaining a message from the intermediary, "processing" is the
+process of acting on the information a message contains, "settling" is the
+process of notifying an intermediary that a message was processed successfully.
+
+### Intermediary
+
+An "intermediary" receives a message to forward it to the next receiver, which
+might be another intermediary or a consumer.
+
+## Scenarios
+
+Producing and consuming a message involves five stages:
+
+```
+PRODUCER
+
+Create
+  |                            CONSUMER
+  v        +--------------+                   
+Publish -> | INTERMEDIARY | -> Receive
+           +--------------+       |
+                  ^               v
+                  .            Process
+                  .               |
+                  .               v
+                  . . . . . .  Settle
+```
+
+1. The producer creates a message.
```

Comment:
> The Create step is left out in current conventions. Since no networking is involved in creation of the message (only in publishing), I did not consider it of interest for messaging semantic convention.

This OTEP is not about proposing an alternative to the existing conventions, but just to clearly establish a shared conceptual model and a set of scenarios we want to cover. We might still decide to model the "Create" and "Publish" steps by a single span, or to not have a span for the "Create" stage at all.

I think it makes sense to have the "Create" and "Publish" stages separate in a conceptual model. 

> From the WG meeting it sounded like Create is actually supposed to mean "logical publish operation" (potentially with retries) while Publish is the "physical publish operation".

That's my understanding, seems we have to make this clearer in the document. For each message, "Create" can only happen once, whereas "Publish" can happen multiple times (due to retries). "Create" (as well as "Process") are agnostic of the underlying protocol (AMQP, MQTT, HTTP), whereas all the other stages are not.

> That is an interesting topic but should maybe be left out initially, as this is already being discussed both generally (#1306) and for HTTP (#1747)

100 % agreed, whatever we come up with must be consistent with HTTP and other similar areas.

Reply:
> Since no networking is involved in creation of the message (only in publishing), I did not consider it of interest for messaging semantic convention

I want to also point that sometimes publishing is performed in batches. there is a single publish operation (span) that can contain multiple messages. For example, SQS supports this [send-message-batch operation](https://awscli.amazonaws.com/v2/documentation/api/latest/reference/sqs/send-message-batch.html)

By excluding a dedicated span for "per message" creation, the following issues might come up:

## `Message Attributes`
Some of the collected attributes are relevant only for the context of a message. for example:
- message id
- conversation id
- instrumentations that allow extended capabilities of capturing payloads and metadata

Attaching these attributes to the publish batch operation will require to set the values as arrays, or exclude them altogether. This behaviour can be problematic at least in these cases:
- confusing to end users and backend processing pipelines, which should be prepared for attributes arriving as either single value (single message publish), or array value (batch publish), and correlate the array indices from different attributes to get a complete set for a single message. If displayed in a raw format (for example, in jaeger), it can be visually hard to observe these long attribute arrays and exract insights from them.
- not symmetric with the "consuming" part which does collect a "per message" span. Although it is possible to handle this asymmetry in code, it does produce a more complex and hard to understand flows.

## `Processing Links`
At least for current conventions, the process span has a span link to the publisher. If the publisher is for batch of messages, then multiple processors for different messages will link to the same span, which again can cause very complex and non-trivial processing of traces.

## `Parietal Failure`
When doing batch publish, the status of the response can be either failure for the entire operation (for example, bad credentials, or network issue), or we can get "success" for some messages and "failure" for others. For example, in SQS, if you do a batch publish of 10 messages, and one message is above the payload limit, the response will contain `Successful` array with 9 message ids, and `Failed` array with 1 message id.
If we don't have the create "per-message" reference, it becomes less straight forward to record this information into publish single-span attributes. I would expect the semantic conventions to support the correct representation of this usecase which I believe best describe when you can reference each message with its own span.

These are real-world problems that I encountered while implementing and processing messaging systems traces.

Reply:
Thanks @blumamir, I agree. I put this on the agenda for tomorrow's meeting, so we can close on this.

Reply:
After thinking on this, and some discussions in today's WG meeting, I would agree it makes sense to separate the Create and Publish stages.

Create is a distinguishable step as there is the possibility to batch many Create steps (and thus messages) into a single Publish step

Reply:
Here's what we discussed today at the meeting:

Separation is needed for:
- batch send (we need unique context per message to trace them individually)
- creating messages, buffering, and publishing in the background as an optimization
- retries on publish (we should also agree on how they are represented)

Should we merge create/publish when there is one message?
- retries?
- one scenario (separate create/publish) is easier to implement and build UX for  

We'll need to define how this separate Create stage looks like (is it a span? if it's not, how we can generate and record a unique context without span) if we agree on the direction.

@Oberon00 does it make sense for you? Do you want to discuss it in more detail?

Reply:
Closing this conversation, because I think there's agreement that it's helpful to have the create/publish stages separate on a conceptual level.

We will have further discussions (probably based on another OTEP) about whether and how we want to reflect that on span creation.